### ISSUE TITLE
Blocking Issue: System.Text.Json and non-public members

### ISSUE THREAD

**Issue by sujayvsarma** [dotnet/runtime#31511]
Hello,

Why do the Serializers and Deserializers in the System.Text.Json namespace not want to deal with non-public class members? Often times, we need to have non-public class properties/fields serialized or deserialized (or even one-way serialized/deserialized). 

This is a blocking issue because Newtonsoft Json does this correctly. This makes porting code that used Newtonsoft to the System.Text.Json serializers impossible.

**Comment by JandosHk** [comment]
EF Core 3.0+ materializes entities directly to backing fields. This is to avoid accidentally executing business logic in property setters, which they may have. STJ must do the same.

**Comment by JandosHk** [comment]
EF Core 3.0+ materializes entities directly to backing fields. This is to avoid accidentally executing business logic in property setters, which they may have. STJ must do the same.

**Comment by layomia** [comment]
Linking https://github.com/dotnet/runtime/issues/38327.

**Comment by layomia** [comment]
Triage - we didn't get to this in .NET 6.0 but can consider it in .NET 7.0.

**Comment by layomia** [comment]
From @swtrse in https://github.com/dotnet/runtime/issues/38327:

> In general, a constructor decorated with the [JsonConstructor] attribute should be able to be a non-public constructor.
> 
> This would address cases where there is a special internal constructor that is solely used for json deserializing and that should not be called from user code.
> 
> Category: Design

**Comment by TanvirArjel** [comment]
.NET 6.0 is going to be most frustrated and useless release since dotNET Core 1.0. Every blocking issue we were following to be fixed in dotNET 6.0 has been moved to the dotNET 7.0!!

RIP dotNET 6.0! 😥 Good Bye. 😀

**Comment by danmoseley** [comment]
@TanvirArjel what were the other issues you were following? What are the top priorities for you?

**Comment by TanvirArjel** [comment]
@daddyman Here is another one (https://github.com/dotnet/runtime/issues/31094) that was supposed to be fixed in .NET 6.0 but .NET Team silently removed this from .NET 6.0 milestone and it seems .NET Team will persist with this hugely misdesigned issue as long as they can.

**Comment by layomia** [comment]
@tanvirarjel it is not a good idea to override the default options used by the serializer, because that could affect serialization in assemblies outside of your own. This may cause serious issues in user apps such as accidentally including or excluding type data. It could be very hard to diagnose such issues.

If you want to have consistent options in your code and wish to not specify them for each (de)serialization call, you could define a wrapper around your serializer calls which pass in the same set of options. Here's an example, adapted from https://github.com/dotnet/runtime/issues/31094. It could also be static, depending on your scenario:

```cs
public class DefaultJsonSerializer
{
    private JsonSerializerOptions _defaultOptions;

    public DefaultJsonSerializer(IOptions<JsonSerializerOptions> defaultOptions)
    {
        _defaultOptions = defaultOptions.Value;
    }

    public async Task<T> DeserializeAsync<T>(Stream utf8Json, JsonSerializerOptions options = default, CancellationToken cancellationToken = default)
        => await JsonSerializer.DeserializeAsync<T>(utf8Json, options ?? _defaultOptions, cancellationToken);
}
```

---

On the default options chosen for the serializer and ASP.NET scenarios: I believe that defaults are opinionated & it is hard to pick values that work well for everyone. Keep in mind that System.Text.Json is a lower level set of APIs and is not always used in web contexts. We've made sure that the options are fully configurable to satisfy a wide range of scenarios.

---

Please take some time to see the new features we've implemented in System.Text.Json in .NET 6.0. There's tons of new functionality in there that many folks will be happy about. There were some things we couldn't get to but we're planning for them in the next release.

---

- https://github.com/dotnet/runtime/issues/43620
- https://github.com/dotnet/runtime/issues/1568
- https://github.com/dotnet/runtime/issues/45441

**Comment by TanvirArjel** [comment]
@layomia If you read all the comments on the issue regarding JSON serializer default  and case insensitive serialization,  you will see that all suggestions and arguments provided by the dotNET team are gracefully rejected by the community. Nobody satisfied with the JSON Serializer design. Everybody wanted breaking change in a major version. 

You have made a high performance JSON serializer, that's fine but at first it has to be workable then performance will come to the party.

I was thrilled when dotNET team announced STJ but now I am frustrated. It's still not ready to replace Newtonsoft.JSON. It also looks weird to use two Serializers in one project.

Thank you.

**Comment by TanvirArjel** [comment]
@layomia  and @danmoseley However, this issue should be fixed in .NET 6.0 as this is a blocking issue and there is no workaround.

**Comment by TanvirArjel** [comment]
@eiriktsarpalis Yes! You are correct. These two enhancements will solve the issues.

**Comment by Romfos** [comment]
Hello, what about 'internal' properties?

**Comment by DrkWzrd** [comment]
In advance, sorry for my english.

The question here is very simple. You want to create a "performance oriented, fast, and easy to use json (de)serialization library", and, at the end, displace Json.NET (don't negate it). Ok, it's fine, it's doable and it's worthy. **_Can we or will we be able to do this in STJ?_**

```C#
    internal class P
    {
        [JsonProperty("_value")]
        private readonly int _value;

        [JsonProperty("_names")]
        private readonly List<string> _names;

        [JsonIgnore]
        public IReadOnlyList<string> Names => _names;

        [JsonConstructor]
        private P(int value)
        {
            _value = value;
            _names = new List<string>();
        }

        public void MakeField()
        {
            Console.WriteLine(_value);
        }

        public void AddName(string name)
        {
            _names.Add(name);
        }

        public static P GetP()
        {
            return new P(Random.Shared.Next());
        }
    }

    static class Program
    {
        static void Main()
        {
            var p = P.GetP();
            p.AddName("santa");
            p.AddName("claus");

            var json = JsonConvert.SerializeObject(p);
            Console.WriteLine(json);

            var p2 = JsonConvert.DeserializeObject<P>(json);
            //everything deserializes like magic!
        }
    }
```
This is a simple, unrealistic, and stupid example, but IT IS on the point.

If the answer is "no", you'll never gonna reach the aceptation level of Json.NET. Moreover, you only gonna frustrate the people who wants to use your API and avoid external references, because "it is easy if I add a simple nuget package".

Sorry if I'm being rude, I don't want to look like that. It's only this could be great, and there are so many little things...

_**I am willing to use STJ**_, but I have, for example, a lot of classes with private readonly collection fields exposed by the IReadOnly- interface by property. And STJ is NOT helping, or the docs are NOT helping (or maybe I'm so stupid I cannot make it, or maybe my POCOs are bad designed).
And yes, I could do a converter, but then all the speed and easiness goes to... nevermind.

At the end of the day the API users will prefer functions over performance.

In the eye of a skeptic person this could be seen as "or you code and (de)serialize my way or GTFO".

**Comment by eiriktsarpalis** [comment]
Hi @DrkWzrd, per your question

> Can we or will we be able to do this in STJ?

Not in .NET 6, at least not without writing your own custom converter that serializes non-public fields.

> If the answer is "no", you'll never gonna reach the aceptation level of Json.NET. Moreover, you only gonna frustrate the people who wants to use your API and avoid external references, because "it is easy if I add a simple nuget package".

Being fully compatible with Json.NET is an explicit non-goal for System.Text.Json. Full feature parity is never going to happen, for many reasons, including security and keeping the product simple and maintainable. We do want to make System.Text.Json _extensible enough_ so that users can plug in their own custom components, but it is never going to be a drop-in replacement for Json.NET. I should add that we thoroughly encourage users taking external NuGet references, BCL never was and never can be a replacement for .NET's rich package ecosystem.

**Comment by TanvirArjel** [comment]
@eiriktsarpalis Any update on this issue. Already two previews of .NET 7.0 have been released.

**Comment by eiriktsarpalis** [comment]
@TanvirArjel to be perfectly honest this is not being prioritized. We have a substantial backlog so it's unlikely we will get round to every feature originally considered for the 7.0.0 milestone. Please refer to the [planned work epic](https://github.com/dotnet/runtime/issues/63762) for a curated list of what we currently have on our roadmap.

**Comment by TanvirArjel** [comment]
@eiriktsarpalis This is a blocking issue and there is no workaround for this in `System.Text.JSON`. Blocking issues should get the highest priority than enhancement, isn't it?

**Comment by Clockwork-Muse** [comment]
> @eiriktsarpalis This is a blocking issue and there is no workaround for this in `System.Text.JSON`. Blocking issues should get the highest priority than enhancement, isn't it?

@TanvirArjel - There's a trivial workaround to make the properties serialized - make them `public` (registering converters and other trick are probably also possible).  This is an enhancement request to _add a feature_ - and one that may not be commonly used, given most cases are going to be using public properties.

**Comment by WhitWaldo** [comment]
Just to jump in with some additional feedback on why I'd like to see this myself. We primarily use JSON to store data in intermediate serialized storage of POCOs and occasionally to simplify communication between services.

We primarily rely on guiding developers on the "right" way to instantiate a given class via the constructors. If the developer can only use these four constructors to instantiate that class coupled with some init setters, we can save ourselves a lot of downstream business logic handling the possibility that properties weren't populated as anticipated.

By making a public parameterless constructor available, even if we mark in the documentation that it's not suitable for developer consumption, it's now there in the Intellisense options and we're relying on the developers having internal object knowledge in order to set the fields properly.

Rather, as we did in Json.NET, I'd really like to see the return of a private parameterless constructor be used for deserialization. It's private, so developers cannot use it themselves (without reflection, which they're not going to use), but it's there just for System.Text.Json to utilize for this purpose.

For me at least, if the .NET 7 STJ release came with nothing but this one ask, I'd be thrilled about it.

Thank you for your continued consideration.

**Comment by OskarKlintrot** [comment]
> it's now there in the Intellisense options
To hide it from intellisense you can use [`[EditorBrowsable(EditorBrowsableState.Never)]`](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.editorbrowsableattribute). Not a perfect solution but it's handy sometimes. It will only hide it though if it's a package you consume.

**Comment by WhitWaldo** [comment]
I've looked for an attribute like that to use as a workaround in the meantime. Thank you @OskarKlintrot !

**Comment by brianlagunas** [comment]
Support for private/internal constructors and properties/fields are also required for my scenario. Please implement this.

**Comment by eiriktsarpalis** [comment]
.NET 7 Preview 6 will include support for [contract customisation](https://github.com/dotnet/runtime/issues/63686), which should enable support for private field serialization for those who need it. Here is a functional demonstrating how this can be done:

https://github.com/dotnet/runtime/blob/5877e8b713742b6d80bd1aa9819094be029e3e1f/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/TypeInfoResolverFunctionalTests.cs#L624-L680

**Comment by TanvirArjel** [comment]
@eiriktsarpalis Will `FieldContractResolver` work for a private constructor too?

**Comment by eiriktsarpalis** [comment]
It should be possible to use private parameterless constructors using the JsonTypeInfo.CreateObject delegate. 

Parameterized constructor metadata has been cut from .NET 7, so it won't be possible to define custom ones.

**Comment by Romfos** [comment]
Thank you for workaround, but [JsonInclude] support would be nice for private properties :(

**Comment by TanvirArjel** [comment]
@eiriktsarpalis Thanks but when will `[JsonInclude]` and `[JsonConstructor]` support for the private properties and constructors arrive?

**Comment by eiriktsarpalis** [comment]
TBH, it is unlikely we would have the time to work on this for .NET 7 at this stage.

**Comment by msynk** [comment]
I think if you make the type itself non-public (internal for example) and change all of its properties to public, somehow the problem can be solved (if you want to hide the type itself and not some of its members ofc).

**Comment by eiriktsarpalis** [comment]
Regarding support for internal member serialization, the release of the [contract customization](https://github.com/dotnet/runtime/issues/63686) feature in .NET 7 Preview 6 should make it possible to achieve via a custom contract resolver. Here is a functional test demonstrating how it could be implemented:

https://github.com/dotnet/runtime/blob/19588127b93f056eb26616d9d31936df94a9e6d8/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/TypeInfoResolverFunctionalTests.cs#L776-L832

**Comment by eiriktsarpalis** [comment]
Per my earlier comment, adding support for non-public members should be achievable via contract customization starting in .NET 7. We're not planning on adding out-of-the-box support for this, but it should still be possible for third-party extension libraries to ship contract resolvers offering this functionality.

**Comment by WhitWaldo** [comment]
@eiriktsarpalis Given that other threads regarding private constructions were closed in favor of this one, can you speak to whether or not support for private constructors is a possibility on any post-.NET 7 roadmaps?

**Comment by TanvirArjel** [comment]
@eiriktsarpalis 

Here is a sample employee class:
```C#
public class Employee
{
    private Employee()
    {
    }

    private Employee(int id, string name)
    {
    }

    public int Id { get; set; }

    public string Name { get; set; }

    public static Employee Create(int id, string name)
    {
        Employee employee = new Employee(id, name);
        return employee;
    }
}

```

Here is my serialization and deserialization code:

```C#
private static void Main(string[] args)
{
    JsonSerializerOptions options = new JsonSerializerOptions 
    { 
      TypeInfoResolver = new PrivateonstructorContractResolver() 
    };

    Employee employee = Employee.Create(1, "Tanvir");
    string json = JsonSerializer.Serialize(employee);
    Employee desrializedEmployee = JsonSerializer.Deserialize<Employee>(json, options);
}
```

Now please write the code for the `PrivateonstructorContractResolver` so that deserialization works:

```C#
public class PrivateonstructorContractResolver : DefaultJsonTypeInfoResolver
{

}
```

**Comment by TanvirArjel** [comment]
@eiriktsarpalis I am eagerly waiting for your response.

**Comment by eiriktsarpalis** [comment]
> @eiriktsarpalis Given that other threads regarding private constructions were closed in favor of this one, can you speak to whether or not support for private constructors is a possibility on any post-.NET 7 roadmaps?

In .NET 7 it should be possible to specify private parameterless constructors via contract customization using the [`JsonTypeInfo.CreateObject`](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.serialization.metadata.jsontypeinfo.createobject?view=net-7.0) delegate. [This issue](https://github.com/dotnet/runtime/issues/71944) tracks parameterized delegate support via contract customization (which should also unblock private constructor support).

> Please write the code for the PrivateConstructorContractResolver so that the above deserialization works:

Per my comments above, parameterized constructor support is not supported yet via the contract customization layer. It should still be possible to get your example to work using the parameterless constructor:

```C#
public class PrivateConstructorContractResolver : DefaultJsonTypeInfoResolver
{
    public override JsonTypeInfo GetTypeInfo(Type type, JsonSerializerOptions options)
    {
          JsonTypeInfo typeInfo = base.GetTypeInfo(type, options);
          if (typeInfo.Type == typeof(Employee))
          {
                typeInfo.CreateObject = () => Employee.Create(); // alternative factory calling into the parameterless ctor
          }

          return typeInfo;
    }
}
```

**Comment by TanvirArjel** [comment]
@eiriktsarpalis What you have written here is tightly coupled to the Employee type so that I have to write contract resolver for every type. Secondly, if i can expose an public parameterless factory method then what's wrong to have public parameterless constructor? I cannot have public parameterless constructor. Please write the contract resolver in such a  way that it works for every type.

**Comment by TanvirArjel** [comment]
@eiriktsarpalis What you have written here is tightly coupled to the Employee type so that I have to write contract resolver for every type. Secondly, if i can expose an public parameterless factory method then what's wrong to have public parameterless constructor? I cannot have public parameterless constructor. Please write the contract resolver in such a  way that it works for every type.

**Comment by eiriktsarpalis** [comment]
Note that a blanket assignment to `CreateObject` might invalidate otherwise valid JSON contracts. I would recommend scoping to the specific types you are interested in customizing, or perhaps use reflection to identify types that don't contain public constructors:
```C#
if (jsonTypeInfo.Kind == JsonTypeInfoKind.Object && jsonTypeInfo.CreateObject is null)
{
     // CreateObject can be null either because we have no constructor or because a parameterized constructor is being used internally
     if (jsonTypeInfo.Type.GetConstructors(BindingFlags.Public | BindingFlags.Instance).Length == 0)
     {
          // The type doesn't have public constructors
          jsonTypeInfo.CreateObject = () => Activator.CreateInstance(jsonTypeInfo.Type, true);
     }
}

**Comment by Clockwork-Muse** [comment]
.... Or writing a source generator to generate per-object registrations, especially since that would enable you to sidestep the need to call `Activator.CreateInstance(...)`.

**Comment by eiriktsarpalis** [comment]
FWIW the .NET 8 release will include support for opting into specific non-public members via the `JsonIncludeAttribute`:

https://github.com/dotnet/runtime/pull/88452

Obviously this still not permit blanket inclusion of non-public members, that can only be achieved using contract customization.
